### ISSUE TITLE
React ApolloProvider component is undefined

### ISSUE THREAD

**Issue by marvinhagemeister** [rollup/rollup#866]
I'm building a react app with [apollo](https://github.com/apollostack/react-apollo) which exports a Provider component similar to the one found in [react-redux](https://github.com/reactjs/react-redux). When bundling with rollup this component is `undefined`.

Repository to reproduce this bug: [https://github.com/marvinhagemeister/rollup-apollo](https://github.com/marvinhagemeister/rollup-apollo)

This is the compiled output of rollup:
```js
var ApolloProvider$1 = createCommonjsModule(function (module, exports) {
"use strict";
var __extends = (commonjsGlobal && commonjsGlobal.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
// ...
var ApolloProvider = (function (_super) {
    __extends(ApolloProvider, _super);
    
    function ApolloProvider(props, context) {
        _super.call(this, props, context);
        // ...
    }
    ApolloProvider.prototype.getChildContext = function () {
        // ...
    }
    // ...
    return ApolloProvider;
}(react_1.Component));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ApolloProvider;
;
});

//
// XXX: this is weird: console.log(ApolloProvider) returns undefined here
//

var ApolloProvider$2 = interopDefault(ApolloProvider$1);

var require$$1$27 = Object.freeze({
    default: ApolloProvider$2
});

// ...

ReactDOM.render(React.createElement(
  ApolloProvider, // <-- fails here, because it is undefined at this point
  { client: client, store: store$10 },
  // ...
), document.getElementById('root'));
```

I'm unsure wether this is an issue in [rollup-plugin-commonjs](https://github.com/rollup/rollup-plugin-commonjs) or rollup.

**Comment by TrySound** [comment]
@marvinhagemeister And where's typescript?

**Comment by marvinhagemeister** [comment]
@TrySound well that was confusing of me. At first I thought it had something to do with ApolloClient itself which is written in typescript. Changed the title to be more clear.

**Comment by TrySound** [comment]
Heh. Without apollo provider everything works :)

**Comment by marvinhagemeister** [comment]
Yep, if I change `ApolloProvider` to `ApolloProvider$2` in the bundle, it everything works fine as well. There is something fishy going on with the module resolution.

```js
ReactDOM.render(React.createElement(
  ApolloProvider$2, // <-- change this,
  // ...
), document.getElementById('root'));
```

**Comment by marvinhagemeister** [comment]
I did some further debugging and narrowed it down to the way the exports of `react-apollo` are declared, which is written in typescript. Seems like my initial assumption wasn't far off after all.

## The Problem

The compiled output of index.js of the react-apollo package expects the `require` call to return an `object`.

```js
var ApolloProvider_1 = require('./ApolloProvider');
exports.ApolloProvider = ApolloProvider_1.default;
```

But `ApolloProvider.js` does indeed export an object:

```js
var ApolloProvider = (function (_super) {
  // ...
});

Object.defineProperty(exports, "__esModule", { value: true });
exports.default = ApolloProvider;
```

So what happens here is that rollup (more likely the commonjs plugin) sees a `default` export and inlines it. 

```js
var ApolloProvider_1 = require('./ApolloProvider');
console.log(ApolloProvider_1); // returns a function, same as default export
console.log(ApolloProvider_1.default); // undefined

exports.ApolloProvider = ApolloProvider_1.default;
```

Inside the bundle `ApolloProvider_1` is **not** an `object` anymore.

**Comment by Rich-Harris** [comment]
The issue here is with rollup-plugin-commonjs. When you do this...

```js
import foo from './commonjs/foo.js';
```

...`foo.js` is massaged such that if it's the output of a transpiler (i.e. instead of using `module.exports` it has `exports.__esModule = true` and `exports.default = whatever`), everything Just Works™ inside the importing module.

But that fails if the importing module is a CommonJS module that *expects* to have to use the `.default` property.

After going round in circles for a bit, I landed on https://github.com/rollup/rollup-plugin-commonjs/pull/92 – thoughts welcome.

**Comment by Rich-Harris** [comment]
I've just released rollup-plugin-commonjs 4.0.0 – could you try it out and see if it resolves this issue? Thanks

**Comment by marvinhagemeister** [comment]
@Rich-Harris Awesome! Just upgraded to the new version and everything is working now. Thanks for having a go at this 👍
